In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv(r"C:\Users\Ali\Desktop\Data Analysis Projects\stack-overflow-developer-survey-2023\survey_results_public.csv")

In [3]:
df.head()

ResponseId     Q120                      MainBranch              Age  \
0           1  I agree                   None of these  18-24 years old   
1           2  I agree  I am a developer by profession  25-34 years old   
2           3  I agree  I am a developer by profession  45-54 years old   
3           4  I agree  I am a developer by profession  25-34 years old   
4           5  I agree  I am a developer by profession  25-34 years old   

                                          Employment  \
0                                                NaN   
1                                Employed, full-time   
2                                Employed, full-time   
3                                Employed, full-time   
4  Employed, full-time;Independent contractor, fr...   

                             RemoteWork  \
0                                   NaN   
1                                Remote   
2  Hybrid (some remote, some in-person)   
3  Hybrid (some remote, some in-person)   
4                                Remote   

                                    CodingActivities  \
0                                                NaN   
1  Hobby;Contribute to open-source projects;Boots...   
2  Hobby;Professional development or self-paced l...   
3                                              Hobby   
4  Hobby;Contribute to open-source projects;Profe...   

                                        EdLevel  \
0                                           NaN   
1  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
2  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
3  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   
4  Bachelor’s degree (B.A., B.S., B.Eng., etc.)   

                                           LearnCode  \
0                                                NaN   
1  Books / Physical media;Colleague;Friend or fam...   
2  Books / Physical media;Colleague;On the job tr...   
3  Colleague;Friend or family member;Other online...   
4  Books / Physical media;Online Courses or Certi...   

                                     LearnCodeOnline  ...        Frequency_1  \
0                                                NaN  ...                NaN   
1  Formal documentation provided by the owner of ...  ...   1-2 times a week   
2  Formal documentation provided by the owner of ...  ...  6-10 times a week   
3  Formal documentation provided by the owner of ...  ...   1-2 times a week   
4  Formal documentation provided by the owner of ...  ...   1-2 times a week   

         Frequency_2       Frequency_3         TimeSearching  \
0                NaN               NaN                   NaN   
1   10+ times a week             Never   15-30 minutes a day   
2  6-10 times a week  3-5 times a week   30-60 minutes a day   
3   10+ times a week  1-2 times a week   15-30 minutes a day   
4   1-2 times a week  3-5 times a week  60-120 minutes a day   

         TimeAnswering                                   ProfessionalTech  \
0                  NaN                                                NaN   
1  15-30 minutes a day  DevOps function;Microservices;Automated testin...   
2  30-60 minutes a day  DevOps function;Microservices;Automated testin...   
3  30-60 minutes a day  Automated testing;Continuous integration (CI) ...   
4  30-60 minutes a day  Microservices;Automated testing;Observability ...   

                                            Industry           SurveyLength  \
0                                                NaN                    NaN   
1  Information Services, IT, Software Development...  Appropriate in length   
2  Information Services, IT, Software Development...  Appropriate in length   
3                                                NaN  Appropriate in length   
4                                              Other  Appropriate in length   

                   SurveyEase ConvertedCompYearly  
0                         NaN                 NaN  
1                        Easy            285000.0  
2              

In [4]:
df.isnull().sum(axis=0)

ResponseId                 0
Q120                       0
MainBranch                 0
Age                        0
Employment              1286
                       ...  
ProfessionalTech       47401
Industry               52410
SurveyLength            2699
SurveyEase              2630
ConvertedCompYearly    41165
Length: 84, dtype: int64

In [5]:
df.columns

Index(['ResponseId', 'Q120', 'MainBranch', 'Age', 'Employment', 'RemoteWork',
       'CodingActivities', 'EdLevel', 'LearnCode', 'LearnCodeOnline',
       'LearnCodeCoursesCert', 'YearsCode', 'YearsCodePro', 'DevType',
       'OrgSize', 'PurchaseInfluence', 'TechList', 'BuyNewTool', 'Country',
       'Currency', 'CompTotal', 'LanguageHaveWorkedWith',
       'LanguageWantToWorkWith', 'DatabaseHaveWorkedWith',
       'DatabaseWantToWorkWith', 'PlatformHaveWorkedWith',
       'PlatformWantToWorkWith', 'WebframeHaveWorkedWith',
       'WebframeWantToWorkWith', 'MiscTechHaveWorkedWith',
       'MiscTechWantToWorkWith', 'ToolsTechHaveWorkedWith',
       'ToolsTechWantToWorkWith', 'NEWCollabToolsHaveWorkedWith',
       'NEWCollabToolsWantToWorkWith', 'OpSysPersonal use',
       'OpSysProfessional use', 'OfficeStackAsyncHaveWorkedWith',
       'OfficeStackAsyncWantToWorkWith', 'OfficeStackSyncHaveWorkedWith',
       'OfficeStackSyncWantToWorkWith', 'AISearchHaveWorkedWith',
       'AISearchWan

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89184 entries, 0 to 89183
Data columns (total 84 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   ResponseId                           89184 non-null  int64  
 1   Q120                                 89184 non-null  object 
 2   MainBranch                           89184 non-null  object 
 3   Age                                  89184 non-null  object 
 4   Employment                           87898 non-null  object 
 5   RemoteWork                           73810 non-null  object 
 6   CodingActivities                     73764 non-null  object 
 7   EdLevel                              87973 non-null  object 
 8   LearnCode                            87663 non-null  object 
 9   LearnCodeOnline                      70084 non-null  object 
 10  LearnCodeCoursesCert                 37076 non-null  object 
 11  YearsCode                   

# 1. How many respondents completed the survey?

In [7]:
df.shape

(89184, 84)

# 2. How many respondents answered all questions?

To determine how many respondents answered all questions, we can check the number of non-null values for each column. If a respondent answered all questions, there should be no missing/null values for that respondent across all columns.

We can sum up the non-null counts for each column, and if that sum equals the total number of rows (89,184) multiplied by the number of columns (84), then all respondents answered all questions.

Here's how we can check that:

In [8]:
total_non_null = df.notnull().sum().sum()
total_cells = df.shape[0] * df.shape[1]

if total_non_null == total_cells:
    print("All respondents answered all questions")
else:
    print(f"Not all respondents answered all questions. Total non-null values: {total_non_null}, Total cells: {total_cells}")

Not all respondents answered all questions. Total non-null values: 4975384, Total cells: 7491456


This means that no single respondent answered all questions, since the total number of non-null values (4975384) is less than the total number of cells (89184 respondents * 84 columns = 7491456 cells).

If all respondents had answered all questions, the total non-null values would be equal to the total number of cells.

Therefore, the answer is that zero (0) respondents answered all questions in this survey.

# 3. What are the values of measures of central tendency for respondents' experience?

To calculate the measures of central tendency (mean, median, mode) for respondents' experience, we need to look at the 'WorkExp' column in the data.

First, let's check for any null/missing values in that column:

In [9]:
df['WorkExp'].isnull().sum()

45605

This gives an output of 45605, meaning there are 45605 null/missing values in the 'WorkExp' column out of the 89184 total respondents.

Next, we'll drop those null values and calculate the mean, median and mode on the remaining non-null values:

In [10]:
work_exp = df['WorkExp'].dropna()

mean_exp = work_exp.mean()
median_exp = work_exp.median()
mode_exp = work_exp.mode()[0]

print(f"Mean experience: {mean_exp:.2f} years")
print(f"Median experience: {median_exp:.2f} years") 
print(f"Mode experience: {mode_exp:.2f} years")

Mean experience: 11.41 years
Median experience: 9.00 years
Mode experience: 5.00 years


# 4. How many respondents work remotely?

To find out how many respondents work remotely, we need to examine the 'RemoteWork' column in the DataFrame.

First, let's check for any null/missing values in that column:

In [11]:
df['RemoteWork'].isnull().sum()

15374

This gives an output of 15374, meaning there are 15374 null/missing values in the 'RemoteWork' column.

Next, we can count the number of respondents who selected 'Remote' as their work mode:

In [12]:
remote_count = df['RemoteWork'].value_counts()['Remote']
print(f"Number of respondents working remotely: {remote_count}")

Number of respondents working remotely: 30566


# 5. What percentage of respondents program in Python?

To find the percentage of respondents who program in Python, we need to look at the 'LanguageHaveWorkedWith' column in the DataFrame. This column contains the programming languages that respondents have worked with, with multiple languages separated by semicolons.

Here are the steps:

Check for null values in 'LanguageHaveWorkedWith'

In [13]:
df['LanguageHaveWorkedWith'].isnull().sum()

2044

This shows there are 2044 null values.

Convert the column to a list of lists, splitting by semicolon

In [14]:
# Split languages by semicolon and check if 'Python' is in the list
python_programmers = df['LanguageHaveWorkedWith'].dropna().apply(lambda x: 'Python' in x.split(';'))

# Count True values 
python_programmers = np.sum(python_programmers)

In [15]:
# Calculate percentage
total_respondents = df.shape[0] - df['LanguageHaveWorkedWith'].isnull().sum()
python_percentage = (python_programmers / total_respondents) * 100

print(f"{python_percentage:.2f}% of respondents program in Python")

49.53% of respondents program in Python


# 6. How many respondents learned to program using online courses?

To find out how many respondents learned to program using online courses, we need to look at the 'LearnCodeOnline' column in the DataFrame.

In [16]:
#Check for null values in the 'LearnCodeOnline' column:
df['LearnCodeOnline'].isnull().sum()

19100

In [17]:
#Filter the DataFrame to include only rows where 'LearnCodeOnline' is not null:
online_learners = df[df['LearnCodeOnline'].notnull()]

In [18]:
#Check the unique values in the 'LearnCodeOnline' column for online_learners:
online_learners['LearnCodeOnline'].unique()

array(['Formal documentation provided by the owner of the tech;Blogs with tips and tricks;Books;Recorded coding sessions;How-to videos;Video-based Online Courses;Written-based Online Courses;Auditory material (e.g., podcasts);Online challenges (e.g., daily or weekly coding challenges);Written Tutorials;Click to write Choice 20;Stack Overflow',
       'Formal documentation provided by the owner of the tech;Blogs with tips and tricks;How-to videos;Online challenges (e.g., daily or weekly coding challenges);Written Tutorials;Click to write Choice 20;Stack Overflow',
       'Formal documentation provided by the owner of the tech;Blogs with tips and tricks;Auditory material (e.g., podcasts);Written Tutorials;Stack Overflow;Interactive tutorial',
       ...,
       'Formal documentation provided by the owner of the tech;Blogs with tips and tricks;Recorded coding sessions;How-to videos;Video-based Online Courses;Written-based Online Courses;Stack Overflow;Certification videos;Other (Please sp

In [19]:
# Check for 'Video-based Online Courses'
video_course_learners = online_learners[online_learners['LearnCodeOnline'].str.contains('Video-based Online Courses', na=False)].shape[0]

# Check for 'Written-based Online Courses' 
written_course_learners = online_learners[online_learners['LearnCodeOnline'].str.contains('Written-based Online Courses', na=False)].shape[0]

# Sum the two counts
online_course_learners = video_course_learners + written_course_learners

print(f"{online_course_learners} respondents learned to program using online courses")

59938 respondents learned to program using online courses


# 7. Among respondents who program in Python, grouped by country, what is the average and median amount of compensation in each country?

Filter the DataFrame to include only rows where 'Python' is present in the 'LanguageHaveWorkedWith' column:

In [20]:
python_programmers = df[df['LanguageHaveWorkedWith'].str.contains('Python', na=False)]

Drop any rows with null values in the 'Country' and 'ConvertedCompYearly' columns:

In [21]:
python_programmers = python_programmers[python_programmers['Country'].notnull() & 
                                        python_programmers['ConvertedCompYearly'].notnull()]

Group the data by 'Country' and calculate the mean and median 'ConvertedCompYearly' for each group:

In [22]:
country_stats = python_programmers.groupby('Country')['ConvertedCompYearly'].agg(['mean', 'median'])
print(country_stats)

                                              mean   median
Country                                                    
Afghanistan                             665.000000     48.0
Albania                               28008.600000  11844.0
Algeria                                8336.333333   6586.0
Andorra                               32127.000000  32127.0
Angola                                  662.000000    662.0
...                                            ...      ...
Venezuela, Bolivarian Republic of...  24973.529412  12000.0
Viet Nam                              20191.870370  13401.0
Yemen                                  8373.000000   9000.0
Zambia                                13051.000000   9687.0
Zimbabwe                               5600.000000   6000.0

[158 rows x 2 columns]


In [23]:
country_stats.sort_values('mean', ascending=False)

mean    median
Country                                          
South Africa              530231.710526   40719.0
United States of America  179229.025050  150000.0
Canada                    178611.953280   89593.5
Switzerland               133818.783951  121238.0
Dominica                  132857.000000  132857.0
...                                 ...       ...
Afghanistan                  665.000000      48.0
Angola                       662.000000     662.0
Brunei Darussalam            592.000000     592.0
Sierra Leone                 283.000000     283.0
Lesotho                       13.000000      13.0

[158 rows x 2 columns]

# 8. What levels of education do the 5 respondents with the highest compensation have?

Drop any rows with null values in the 'ConvertedCompYearly' and 'EdLevel' columns:

In [24]:
data = df[df['ConvertedCompYearly'].notnull() & df['EdLevel'].notnull()]

Sort the data by 'ConvertedCompYearly' in descending order to get the highest compensated respondents first:

In [25]:
data = data.sort_values('ConvertedCompYearly', ascending=False)

Select the top 5 rows and display their 'EdLevel' values:

In [26]:
top_5_edu_levels = data['EdLevel'].head(5)
print(top_5_edu_levels)

53268    Professional degree (JD, MD, Ph.D, Ed.D, etc.)
77848    Professional degree (JD, MD, Ph.D, Ed.D, etc.)
66223      Bachelor’s degree (B.A., B.S., B.Eng., etc.)
28121                         Primary/elementary school
19679    Professional degree (JD, MD, Ph.D, Ed.D, etc.)
Name: EdLevel, dtype: object


# 9. In each age category, what percentage of respondents program in Python?

Check for null values in the 'LanguageHaveWorkedWith' and 'Age' columns:

In [27]:
print(f"Null values in 'LanguageHaveWorkedWith': {df['LanguageHaveWorkedWith'].isnull().sum()}")
print(f"Null values in 'Age': {df['Age'].isnull().sum()}")

Null values in 'LanguageHaveWorkedWith': 2044
Null values in 'Age': 0


Drop any rows with null values in these two columns:

In [28]:
data = df.dropna(subset=['LanguageHaveWorkedWith', 'Age'])

Create a boolean mask for Python programmers:

In [29]:
python_programmers = data['LanguageHaveWorkedWith'].str.contains('Python', na=False)

Group the data by 'Age' and calculate the percentage of Python programmers in each age group:

In [30]:
python_percentage = data.groupby('Age')['LanguageHaveWorkedWith'].apply(lambda x: (x[python_programmers].count() / x.count()) * 100)
print(python_percentage)

Age
18-24 years old       63.054187
25-34 years old       48.361257
35-44 years old       42.230384
45-54 years old       39.363793
55-64 years old       37.697929
65 years or older     32.879201
Prefer not to say     57.632399
Under 18 years old    71.234599
Name: LanguageHaveWorkedWith, dtype: float64


This shows that, for example, 63.05% of respondents in the '18-24 years old' age group program in Python.

# 10. Among respondents who are in the 75th percentile for average compensation and work remotely, which industries are the most common?

Drop any rows with null values in the 'RemoteWork', 'ConvertedCompYearly', and 'Industry' columns:

In [31]:
data = df.dropna(subset=['RemoteWork', 'ConvertedCompYearly', 'Industry'])

Filter the data to include only respondents who work remotely:

In [32]:
remote_workers = data[data['RemoteWork'] == 'Remote']

Calculate the 75th percentile of compensation for remote workers:

In [33]:
compensation_75th_percentile = remote_workers['ConvertedCompYearly'].quantile(0.75)

Filter the data to include only respondents above the 75th percentile of compensation:

In [34]:
high_earners = remote_workers[remote_workers['ConvertedCompYearly'] >= compensation_75th_percentile]

Count the occurrences of each industry for the high-earning remote workers:

In [35]:
industry_counts = high_earners['Industry'].value_counts()

Sort the counts in descending order to find the most common industries:

In [36]:
most_common_industries = industry_counts.sort_values(ascending=False)
print(most_common_industries)

Industry
Information Services, IT, Software Development, or other Technology    1425
Financial Services                                                      424
Other                                                                   345
Healthcare                                                              278
Retail and Consumer Services                                            184
Manufacturing, Transportation, or Supply Chain                          120
Advertising Services                                                     92
Insurance                                                                69
Higher Education                                                         47
Oil & Gas                                                                14
Legal Services                                                           14
Wholesale                                                                 6
Name: count, dtype: int64
